kube-runtime/src/wait.rs

```diff
@@ -0,0 +1,95 @@
+use futures::TryStreamExt;
+use kube::{Api, Resource};
+use serde::de::DeserializeOwned;
+use snafu::{futures::TryStreamExt as _, Snafu};
+use std::fmt::Debug;
+
+use crate::watcher::{self, watch_object};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("failed to probe for whether the condition is fulfilled yet: {}", source))]
+    ProbeFailed {
+        #[snafu(backtrace)]
+        source: watcher::Error,
+    },
+}
+
+/// Watch an object, and Wait for some condition `cond` to return `true`.
+///
+/// `cond` is passed `Some` if the object is found, otherwise `None`.
+///
+/// # Errors
+///
+/// Fails if the type is not known to the Kubernetes API, or if the [`Api`] does not have
+/// permission to `watch` and `list` it.
+///
+/// Does *not* fail if the object is not found.
+pub async fn await_condition<K>(
+    api: Api<K>,
+    name: &str,
+    mut cond: impl FnMut(Option<&K>) -> bool,
+) -> Result<(), Error>
+where
+    K: Clone + Debug + Send + DeserializeOwned + Resource + 'static,
+{
+    watch_object(api, name)
+        .context(ProbeFailed)
+        .try_take_while(|obj| {
+            let result = !cond(obj.as_ref());
+            async move { Ok(result) }
+        })
+        .try_for_each(|_| async { Ok(()) })
```

Comment:
I'm not really sure what that would add, over the user just wrapping the `Future` with [`tokio::time::timeout`](https://docs.rs/tokio/1.11.0/tokio/time/fn.timeout.html) themself? Is it "just" to force the user to consider that they might want it?

Reply:
Well, watch is the only api call we hold indefinitely. I think it would make sense to make the timeout use explicit as a good practice.

I.e. make the user thing about it, plus provide a convenience for the common use case. It's not a big add, but if we don't do it then newbie users would be exposed to `select!`, and some will have to reason about whether that's a safe thing to do (e.g. what happens to the other future), which puts a bit of an unnecessary understanding barrier there compared to passing in a `Duration` and catching a `TimeoutError`.

We have to force it on users I think. We could probably optionally select wrap at the end (e.g. if something like `WatchParams::timeout` is a `Some(Duration)`).

Reply:
> It's not a big add, but if we don't do it then newbie users would be exposed to select!, and some will have to reason about whether that's a safe thing to do (e.g. what happens to the other future),

No they wouldn't? `tokio::time::timeout` wraps the `Future` specifically to avoid this. The difference would be between:

```rust
await_condition(api, "foo", bar_cond, WaitParams {
    timeout: Some(Duration::from_secs(10)),
}).await?
```

and

```rust
timeout(Duration::from_secs(10), await_condition(api, "foo", bar_cond)).await??
```

Reply:
Oh what. That is great! I thought for sure it worked the other way.

Also; I'm sorry. seems I am continually wasting your time here by failing to read api docs :(